### DISCUSSION TITLE
M4-Report

### DISCUSSION THREAD

**Discussion by petertheprocess** [sopra-fs24-group-09/sopra-fs24-group-09-client#188] [Announcements]
# Introduction
# Challenges
- Our game requires sharing reversed audio over a network, and we have high demands for high-quality audio compression and rapid audio reversal. Finding a stable and efficient audio processing package to meet our needs is challenging. 
- Each round of the game features a countdown timer that must be synchronized among multiple players, ensuring that one player‘s network latency does not affect the experience for other players.
- The game logic is quite complex for new players, making it challenging for them to get started. A non-interactive text-based guide is insufficient for this purpose. 
- The lobby information includes the number of players in each room and the current game status. Keeping this information synchronized in real-time across all players presents a challenge due to the need for instant updates and accuracy.
- Handling network failures and reconnecting with minimal impact on other players is indeed another challenge. 
# What We Learned
- **Continuous integration is crucial for collaborative software development.** Our trategy is maintain one main branch, one develop branch and many feature brances. The main branch is the version we deployed in the cloud, if we wanna some new feature or do some bug fixing, we create a new feature branch and work on that. Once development on a feature branch is complete, developers create a pull request (PR) to merge into the dev branch. This PR undergoes thorough testing and code review by team members to ensure any potential bugs introduced through team collaboration are caught early. This approach helps maintain code quality and stability before changes are integrated into the dev branch and eventually into the main branch.
- **ESlint in powerful**. Integrating ESLint into our development process has been essential for maintaining code quality. For each pull request, the ESLint check runs automatically. If it fails, the merge is not allowed. It enforces coding standards, catches common errors early, and helps ensure consistent coding practices across the team. Before M3, we ignored many linting rules, leading to difficult-to-understand code during debugging. Enforcing these rules has made our codebase cleaner and more maintainable, reducing the time spent on debugging and increasing overall development efficiency.

**Comment by Zehao-Zhang** [comment]
# Challenges in server end

Given the need to store entities a n d ensure smooth operation in a production environment, maintaining and managing data persistency is a big challenge. To address these issues, we decided to integrate an external database. By leveraging MongoDB, we aim to enhance our data persistence capabilities. And also MongoDB offers high flexibility and performance to our application.

Considering the limited memory in Google cloud, handling concurrency in the cloud is a significant challenge. To prevent potential scaling issues, we create a thread pool to pre-allocate resources. We restrict maximum 10 threads running at the same time, it can prevent high traffic waiting and caused errors to our application.

**Comment by Haaaan1** [comment]
**Challenges:**
Implementing a robust authentication system is challenging. We need to continuously validate the user's token to ensure it is still valid. It is challenging to make sure when the user logs into a new tab, the expired session should be disconnected and cleared. This adds an extra layer of complexity to both front-end and back-end logic and requires careful handling.

**What We Learned**
**Importance of authentication in maintaining security and user session integrity.** By adding tokens to headers for each request and conversation, we ensure user validity. Implementing token checks across HTTP and WebSocket communications helps prevent conflicts, such as handling session expirations when a user logs in from a new window, thereby invalidating the old session and ensuring secure, conflict-free user experiences.

**Comment by Haaaan1** [comment]
**Challenges:**
Implementing user-friendly notifications that persist across page transitions is challenging. Notifications should be visible and informative, ensuring users are aware of actions such as successful logins.

**What We Learned**
**Custom toast component to handle notifications.** By integrating this component into our main application, we ensured that notifications persist across page transitions and are accessible throughout the entire application. This approach provided a consistent and user-friendly experience, allowing important messages to be visible even after navigating to a new page.
